compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -365,14 +366,54 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                 }
 
                 self.infcx.probe(|_| {
-                    if let Ok(_substs) = self.match_impl(impl_def_id, impl_trait_ref, obligation) {
+                    if let Ok(_) = self.match_impl(impl_def_id, impl_trait_ref, obligation) {
+                        if self.reject_fn_ptr_impls(impl_def_id, obligation) {
+                            return;
+                        }
                         candidates.vec.push(ImplCandidate(impl_def_id));
                     }
                 });
             },
         );
     }
 
+    fn reject_fn_ptr_impls(&self, impl_def_id: DefId, obligation: &TraitObligation<'tcx>) -> bool {
+        // Do not consider `impl<T: FnDef> Trait for T` for types other than function pointers
+        // or generic parameters that have a `FnDef` bound in scope.
+        for &(predicate, _) in self.tcx().predicates_of(impl_def_id).predicates {
+            let ty::PredicateKind::Clause(ty::Clause::Trait(pred))
+                = predicate.kind().skip_binder() else { continue };
+            if self.tcx().lang_items().fn_ptr_trait() != Some(pred.trait_ref.def_id) {
+                continue;
+            }
+            match pred.self_ty().kind() {
+                // Fast path to avoid evaluating an obligation that trivally holds.
+                ty::FnDef(..) => continue,
+                ty::Param(..) => {}
+                // No other types can implement `FnPtr`
+                _ => return true,
+            }
+
+            // Generic params can implement `FnPtr` if the predicate
+            // holds within its own environment.
+            let obligation = Obligation::new(
+                self.tcx(),
+                obligation.cause.clone(),
+                obligation.param_env,
+                self.tcx().mk_predicate(obligation.predicate.map_bound(|obl_pred| {
+                    ty::EarlyBinder(predicate.kind().no_bound_vars().unwrap())
+                        .subst(self.tcx(), obl_pred.trait_ref.substs)
```

Comment:
these substs are wrong. You would need to use the `impl_substs` (returned by `match_impl`) here instead of the trait substs: 
```rust
trait Foo {}
impl<T> Foo for Vec<T>
where
    T: FnPtr,
{}
```
if you now had `Vec<fn()>: Foo` we would look at the impl, call substs on `T: FnPtr` with the substs from the trait, which are `Vec<fn()>` instead of the impl substs which would be `fn()`.